diff --git a/os/os.hpp b/os/os.hpp
index 82879de..562a98c 100755
--- a/os/os.hpp
+++ b/os/os.hpp
@@ -39,6 +39,10 @@
 #define MIN_PAGE_SHIFT 12
 #define MIN_PAGE_SIZE (1UL << MIN_PAGE_SHIFT)
 
+#ifndef __cpu_mask
+typedef unsigned long __cpu_mask;
+#endif 
+
 namespace amd {
 
 /*! \addtogroup Os Operating System Abstraction
diff --git a/os/os_posix.cpp b/os/os_posix.cpp
index 91d08bc..e5b8818 100755
--- a/os/os_posix.cpp
+++ b/os/os_posix.cpp
@@ -38,6 +38,7 @@
 #include <pthread.h>
 #include <dlfcn.h>
 #include <signal.h>
+#include <libgen.h>
 
 #include <sys/prctl.h>
 
@@ -374,9 +375,6 @@ const void* Os::createOsThread(amd::Thread* thread) {
     for (int i = 0; i < processorCount_; i++) {
       CPU_SET(i, &cpuset);
     }
-    if (0 != pthread_attr_setaffinity_np(&threadAttr, sizeof(cpu_set_t), &cpuset)) {
-      fatal("pthread_attr_setaffinity_np failed to set affinity");
-    }
   }
 
   pthread_t handle = 0;
