--- src/config.h
+++ src/config.h
@@ -12,9 +12,9 @@
     // const static std::string playcmd_start = "youtube-dl -o - \"";
     // const static std::string playcmd_end = "\" | mpv -";
 
-    const static std::string playcmd_start = "mpv \"";
+    const static std::string playcmd_start = "youtube-dl -o - \"";
     // url is placed in middle
-    const static std::string playcmd_end = "\"";
+    const static std::string playcmd_end = "\" | mpv -";
 }
 
 #endif //TUITUBE_CONFIG_H
