diff -r -u webkit2gtk-orig/Source/JavaScriptCore/runtime/IntlSegmenter.cpp webkit2gtk/Source/JavaScriptCore/runtime/IntlSegmenter.cpp
--- webkit2gtk-orig/Source/JavaScriptCore/runtime/IntlSegmenter.cpp	2021-04-10 10:50:56.502202802 +0300
+++ webkit2gtk/Source/JavaScriptCore/runtime/IntlSegmenter.cpp	2021-04-10 10:51:31.100166211 +0300
@@ -125,7 +125,7 @@
     auto upconvertedCharacters = Box<Vector<UChar>>::create(string.charactersWithoutNullTermination());
 
     UErrorCode status = U_ZERO_ERROR;
-    auto segmenter = std::unique_ptr<UBreakIterator, UBreakIteratorDeleter>(ubrk_safeClone(m_segmenter.get(), nullptr, nullptr, &status));
+    auto segmenter = std::unique_ptr<UBreakIterator, UBreakIteratorDeleter>(ubrk_clone(m_segmenter.get(), &status));
     if (U_FAILURE(status)) {
         throwTypeError(globalObject, scope, "failed to initialize Segments"_s);
         return { };
diff -r -u webkit2gtk-orig/Source/JavaScriptCore/runtime/IntlSegments.cpp webkit2gtk/Source/JavaScriptCore/runtime/IntlSegments.cpp
--- webkit2gtk-orig/Source/JavaScriptCore/runtime/IntlSegments.cpp	2021-04-10 10:50:56.499202805 +0300
+++ webkit2gtk/Source/JavaScriptCore/runtime/IntlSegments.cpp	2021-04-10 10:51:55.974141629 +0300
@@ -100,7 +100,7 @@
     auto scope = DECLARE_THROW_SCOPE(vm);
 
     UErrorCode status = U_ZERO_ERROR;
-    auto segmenter = std::unique_ptr<UBreakIterator, UBreakIteratorDeleter>(ubrk_safeClone(m_segmenter.get(), nullptr, nullptr, &status));
+    auto segmenter = std::unique_ptr<UBreakIterator, UBreakIteratorDeleter>(ubrk_clone(m_segmenter.get(), &status));
     if (U_FAILURE(status)) {
         throwTypeError(globalObject, scope, "failed to initialize SegmentIterator"_s);
         return nullptr;
