--- a/Makefile
+++ b/Makefile
@@ -1,15 +1,5 @@
 pkgconfig = $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) pkg-config $(1))
 
-ifeq ($(shell uname),Darwin)
-  LIBTOOL ?= glibtool
-else
-  LIBTOOL ?= libtool
-endif
-
-ifneq ($(VERBOSE),1)
-  LIBTOOL +=--quiet
-endif
-
 override CFLAGS +=-Wall -std=c99
 
 ifeq ($(DEBUG),1)
@@ -67,31 +57,31 @@
 all: $(LIBRARY) $(DEMOS)
 
 %.lo: %.c termkey.h termkey-internal.h
-	$(LIBTOOL) --mode=compile --tag=CC $(CC) $(CFLAGS) -o $@ -c $<
+	$(CC) $(CFLAGS) -fPIC -o $@ -c $<
 
 $(LIBRARY): $(OBJECTS)
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -rpath $(LIBDIR) -version-info $(VERSION_CURRENT):$(VERSION_REVISION):$(VERSION_AGE) $(LDFLAGS) -o $@ $^
+	$(CC) -shared -Wl,-rpath,$(LIBDIR),$(LDFLAGS) -o $@ $^
 
 libtermkey.a: $(OBJECTS)
-	$(LIBTOOL) --mode=link $(CC) -o $@ $^
+	$(AR) rc $@ $^
 
 demo: $(LIBRARY) demo.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^
+	$(CC) -o $@ $^
 
 demo-async: $(LIBRARY) demo-async.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^
+	$(CC) -o $@ $^
 
 demo-glib.lo: demo-glib.c termkey.h
-	$(LIBTOOL) --mode=compile --tag=CC $(CC) -o $@ -c $< $(call pkgconfig, glib-2.0 --cflags)
+	$(CC) -shared -o $@ -c $< $(call pkgconfig, glib-2.0 --cflags)
 
 demo-glib: $(LIBRARY) demo-glib.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^ $(call pkgconfig, glib-2.0 --libs)
+	$(CC) -o $@ $^ $(call pkgconfig, glib-2.0 --libs)
 
 t/%.t: t/%.c $(LIBRARY) t/taplib.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^
+	$(CC) -o $@ $^
 
 t/taplib.lo: t/taplib.c
-	$(LIBTOOL) --mode=compile --tag=CC $(CC) $(CFLAGS) -o $@ -c $^
+	$(CC) -shared $(CFLAGS) -o $@ -c $^
 
 .PHONY: test
 test: $(TESTFILES)
@@ -99,17 +89,16 @@
 
 .PHONY: clean-test
 clean-test:
-	$(LIBTOOL) --mode=clean rm -f $(TESTFILES) t/taplib.lo
+	rm -f $(TESTFILES) t/taplib.lo
 
 .PHONY: clean
 clean: clean-test
-	$(LIBTOOL) --mode=clean rm -f $(OBJECTS) $(DEMO_OBJECTS)
-	$(LIBTOOL) --mode=clean rm -f $(LIBRARY)
-	$(LIBTOOL) --mode=clean rm -rf $(DEMOS)
+	rm -f $(OBJECTS) $(DEMO_OBJECTS)
+	rm -f $(LIBRARY)
+	rm -rf $(DEMOS)
 
 .PHONY: install
 install: install-inc install-lib install-man
-	$(LIBTOOL) --mode=finish $(DESTDIR)$(LIBDIR)
 
 install-inc: termkey.h
 	install -d $(DESTDIR)$(INCDIR)
@@ -119,8 +108,8 @@
 
 install-lib: $(LIBRARY) libtermkey.a
 	install -d $(DESTDIR)$(LIBDIR)
-	$(LIBTOOL) --mode=install install libtermkey.la $(DESTDIR)$(LIBDIR)/libtermkey.la
-	$(LIBTOOL) --mode=install install libtermkey.a $(DESTDIR)$(LIBDIR)/libtermkey.a
+	install libtermkey.la $(DESTDIR)$(LIBDIR)/libtermkey.la
+	install libtermkey.a $(DESTDIR)$(LIBDIR)/libtermkey.a
 
 install-man:
 	install -d $(DESTDIR)$(MAN3DIR)
