--- a/Makefile
+++ b/Makefile
@@ -35,16 +35,8 @@
 endif
 
 OBJECTS=termkey.lo driver-csi.lo driver-ti.lo
-LIBRARY=libtermkey.la
+LIBRARY=libtermkey.so
 
-DEMOS=demo demo-async
-
-ifeq ($(call pkgconfig, glib-2.0 && echo 1),1)
-  DEMOS+=demo-glib
-endif
-
-DEMO_OBJECTS=$(DEMOS:=.lo)
-
 TESTSOURCES=$(wildcard t/[0-9]*.c)
 TESTFILES=$(TESTSOURCES:.c=.t)
 
@@ -64,49 +56,34 @@
 MAN3DIR=$(MANDIR)/man3
 MAN7DIR=$(MANDIR)/man7
 
-all: $(LIBRARY) $(DEMOS)
+all: $(LIBRARY) libtermkey.a
 
 %.lo: %.c termkey.h termkey-internal.h
-	$(LIBTOOL) --mode=compile --tag=CC $(CC) $(CFLAGS) -o $@ -c $<
+	$(CC) $(CFLAGS) -fPIC -o $@ -c $<
 
 $(LIBRARY): $(OBJECTS)
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -rpath $(LIBDIR) -version-info $(VERSION_CURRENT):$(VERSION_REVISION):$(VERSION_AGE) $(LDFLAGS) -o $@ $^
+	$(CC) -shared -Wl,-rpath,$(LIBDIR) $(LDFLAGS) -o $@ $^
 
-demo: $(LIBRARY) demo.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^
+libtermkey.a: $(OBJECTS)
+	$(AR) rc $@ $^
 
-demo-async: $(LIBRARY) demo-async.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^
-
-demo-glib.lo: demo-glib.c termkey.h
-	$(LIBTOOL) --mode=compile --tag=CC $(CC) -o $@ -c $< $(call pkgconfig, glib-2.0 --cflags)
-
-demo-glib: $(LIBRARY) demo-glib.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^ $(call pkgconfig, glib-2.0 --libs)
-
 t/%.t: t/%.c $(LIBRARY) t/taplib.lo
-	$(LIBTOOL) --mode=link --tag=CC $(CC) -o $@ $^
+	$(CC) -o $@ $^
 
 t/taplib.lo: t/taplib.c
-	$(LIBTOOL) --mode=compile --tag=CC $(CC) $(CFLAGS) -o $@ -c $^
+	$(CFLAGS) -o $@ -c $^
 
 .PHONY: test
 test: $(TESTFILES)
 	prove -e ""
 
-.PHONY: clean-test
-clean-test:
-	$(LIBTOOL) --mode=clean rm -f $(TESTFILES) t/taplib.lo
-
 .PHONY: clean
-clean: clean-test
-	$(LIBTOOL) --mode=clean rm -f $(OBJECTS) $(DEMO_OBJECTS)
-	$(LIBTOOL) --mode=clean rm -f $(LIBRARY)
-	$(LIBTOOL) --mode=clean rm -rf $(DEMOS)
+clean:
+	rm -f $(OBJECTS)
+	rm -f $(LIBRARY)
 
 .PHONY: install
 install: install-inc install-lib install-man
-	$(LIBTOOL) --mode=finish $(DESTDIR)$(LIBDIR)
 
 install-inc: termkey.h
 	install -d $(DESTDIR)$(INCDIR)
@@ -116,7 +93,8 @@
 
 install-lib: $(LIBRARY)
 	install -d $(DESTDIR)$(LIBDIR)
-	$(LIBTOOL) --mode=install install libtermkey.la $(DESTDIR)$(LIBDIR)/libtermkey.la
+	install $(LIBRARY) $(DESTDIR)$(LIBDIR)/$(LIBRARY)
+	install libtermkey.a $(DESTDIR)$(LIBDIR)/libtermkey.a
 
 install-man:
 	install -d $(DESTDIR)$(MAN3DIR)
