diff --git a/kirc.c b/kirc.c
index 4e9cb24..da2a71c 100644
--- a/kirc.c
+++ b/kirc.c
@@ -38,7 +38,7 @@ static int    conn;                      /* connection socket */
 static int    verb = 0;                  /* verbose output (e.g. raw stream) */
 static size_t cmax = 80;                 /* max number of chars per line */
 static size_t gutl = 10;                 /* max char width of left column */
-static char   chan[CHA_MAX] = "kirc";    /* channel */
+static char   chan[CHA_MAX] = "kisslinux"; /* channel */
 static char * host = "irc.freenode.org"; /* irc host address */
 static char * port = "6667";             /* server port */
 static char * nick = NULL;               /* nickname */
