--- linux-user/syscall.c
+++ linux-user/syscall.c
@@ -46,7 +46,7 @@
 #include <sys/sem.h>
 #include <sys/statfs.h>
 #include <utime.h>
-#include <sys/sysinfo.h>
+//#include <sys/sysinfo.h>
 #include <sys/signalfd.h>
 //#include <sys/user.h>
 #include <netinet/ip.h>
@@ -112,6 +112,13 @@
 #include "qapi/error.h"
 #include "fd-trans.h"

+#ifndef F_SHLCK
+#define F_SHLCK 8
+#endif
+#ifndef F_EXLCK
+#define F_EXLCK 4
+#endif
+
 #ifndef CLONE_IO
 #define CLONE_IO                0x80000000      /* Clone io context */
 #endif
@@ -6490,8 +6497,17 @@
     return 0;
 }

+struct host_sigevent {
+    union sigval sigev_value;
+    int sigev_signo;
+    int sigev_notify;
+    union {
+       int _pad[64-sizeof(int) * 2 + sizeof(union sigval)];
+       int _tid;
+    } _sigev_un;
+};

-static inline abi_long target_to_host_sigevent(struct sigevent *host_sevp,
+static inline abi_long target_to_host_sigevent(struct host_sigevent *host_sevp,
                                                abi_ulong target_addr)
 {
     struct target_sigevent *target_sevp;

